pinot-controller/src/main/java/org/apache/pinot/controller/recommender/io/InputManager.java

```diff
@@ -0,0 +1,519 @@
+   */
+  public boolean isDim(String colName) {
+    return _dimNames.contains(colName);
+  }
+
+  public boolean isDateTime(String colName) {
+    return _schema.getTimeFieldSpec().getName().equals(colName);
+  }
+
+  public void registerColnameFieldType() { // create a map from colname to data type
+    for (DimensionFieldSpec dimensionFieldSpec : _schema.getDimensionFieldSpecs()) {
+      _colnameFieldTypeMap.put(dimensionFieldSpec.getName(), dimensionFieldSpec.getDataType());
+    }
+    for (MetricFieldSpec metricFieldSpec : _schema.getMetricFieldSpecs()) {
+      _colnameFieldTypeMap.put(metricFieldSpec.getName(), metricFieldSpec.getDataType());
+    }
+    //TODO: add support for multiple getDateTimeFieldSpecs
+    _colnameFieldTypeMap.put(_schema.getTimeFieldSpec().getName(), _schema.getTimeFieldSpec().getDataType());
+  }
+
+  public void estimateSizePerRecord() {
+    for (String colName : _colnameFieldTypeMap.keySet()) {
+      _sizePerRecord += getColDataSizeWithDictionary(colName);
+      LOGGER.debug("{} {}",colName, getColDataSizeWithDictionary(colName));
+    }
+    LOGGER.info("*Estimated size per record {} bytes", _sizePerRecord);
+  }
+
+  public long getColDataSizeWithoutDictionary(String colName) {
+    //TODO: implement this after the complex is supported
+    FieldSpec.DataType dataType = getFieldType(colName);
+    if (dataType == FieldSpec.DataType.STRUCT || dataType == FieldSpec.DataType.MAP
+        || dataType == FieldSpec.DataType.LIST) {
+      return 0;
+    } else {
+      if (dataType == FieldSpec.DataType.BYTES || dataType == FieldSpec.DataType.STRING) {
+        return _dataTypeSizeMap.get(dataType) * getAverageDataLen(colName);
+      } else {
+        return _dataTypeSizeMap.get(dataType);
+      }
+    }
+  }
+
+  public long getColDataSizeWithDictionary(String colName) {
+    //TODO: implement this after the complex is supported
+    FieldSpec.DataType dataType = getFieldType(colName);
+    int numValuesPerEntry = (int) Math.ceil(getNumValuesPerEntry(colName));
+    LOGGER.trace("{} {}", colName, numValuesPerEntry);
+    if (dataType == FieldSpec.DataType.STRUCT || dataType == FieldSpec.DataType.MAP
+        || dataType == FieldSpec.DataType.LIST) {
+      return 0;
+    } else if (!_overWrittenConfigs.getIndexConfig().getNoDictionaryColumns().contains(colName)) { // has dictionary
+      return getBitCompressedDataSize(colName) * numValuesPerEntry;
+    } else { // no dictionary
+      if (dataType == FieldSpec.DataType.BYTES || dataType == FieldSpec.DataType.STRING) {
+        return _dataTypeSizeMap.get(dataType) * numValuesPerEntry * getAverageDataLen(colName);
+      } else {
+        return _dataTypeSizeMap.get(dataType) * numValuesPerEntry;
+      }
+    }
+  }
+
+  public int getBitCompressedDataSize(String colName) {
+    return max((int) Math.ceil(Math.log(getCardinality(colName)) / (8 * Math.log(2))), 1);
+  }
+
+  //
+  public long getDictionarySize(String colName) {
```

Comment:
This function is only computing the size of dictionary right? We should not include the size of bit compressed forward index

Reply:
Done! Thanks for pointing out this bug.